File: AoC2023/AoC2023/Two/TwoOne.cs
namespace AoC2023.Two;

public static class TwoOne
{
    public static int Run(string dataFilepath)
    {
        var maxValue = 0;
        foreach (var dataRow in File.ReadLines(dataFilepath))
        {
            var game = dataRow.Split(':');
            var rounds = game[1].Split(';');
            var isPossible = true;
            foreach (var round in rounds)
            {
                var scores = round.Trim().Split(' ');
                for (var i = 0; i < scores.Length / 2; i++)
                {
                    var score = scores.Skip(2 * i).Take(2).ToArray();
                    var color = score[1].TrimEnd(',');
                    var number = int.Parse(score[0]);

                    switch (color)
                    {
                        case "red" when number > 12:
                        case "green" when number > 13:
                        case "blue" when number > 14:
                            isPossible = false;
                            break;
                    }
                }
            }
            
            if (isPossible)
            {
                maxValue += int.Parse(game[0].Split(' ')[1]);
            }
        }

        return maxValue;
    }
}

File: AoC2023-GO/TwoOne/twoone.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	file, err := os.Open("../../Data2023/DayTwoData.txt")
	if err != nil {
		fmt.Print(err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	maxValue := 0
	for scanner.Scan() {
		game := strings.Split(scanner.Text(), ":")
		rounds := strings.Split(game[1], ";")
		isPossible := true
		for _, round := range rounds {
			scores := strings.Split(strings.Trim(round, " "), " ")

			for i := 0; i < len(scores)/2; i++ {
				score := scores[2*i : 2+2*i]
				color := strings.TrimSuffix(score[1], ",")
				number, _ := strconv.Atoi(score[0])

				if color == "red" && number > 12 {
					isPossible = false
					break
				}

				if color == "green" && number > 13 {
					isPossible = false
					break
				}

				if color == "blue" && number > 14 {
					isPossible = false
					break
				}
			}
		}

		if isPossible {
			round, _ := strconv.Atoi(strings.Split(game[0], " ")[1])
			maxValue += round
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Print(err)
	}

	fmt.Println(maxValue)
}
